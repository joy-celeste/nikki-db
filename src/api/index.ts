import { HttpRequestException } from '../models/Errors';
import database from '../Database';

export const fetchItemData = async (itemId: number): Promise<any> => {
  try {
    return database.collection('clothes').doc(itemId.toString())
      .get()
      .then((doc) => {
        if (!doc.exists) {
          console.log('No such document!');
          return null;
        }
        console.log('Retrieved data from database.', doc.data());
        return doc.data();
      });
  } catch (error) {
    throw new HttpRequestException(error.code, 'Error getting document.');
  }
};

export default fetchItemData;
