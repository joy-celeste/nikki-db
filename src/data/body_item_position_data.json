{"position":{"30":{"pot_scale":1,"posy":62,"posx":-7.5},"24":{"pot_scale":1,"posy":62,"posx":-9.5},"23":{"pot_scale":1,"posy":257.5,"posx":-18},"22":{"pot_scale":1,"posy":62,"posx":-7.5},"17":{"pot_scale":1,"posy":254,"posx":-23},"25":{"pot_scale":1,"posy":254,"posx":-23},"6":{"posy":299.5,"posx":-23.5},"7":{"posy":376,"posx":-30.5},"8":{"posy":260,"posx":-24},"9":{"posy":146,"posx":-22.5},"29":{"pot_scale":1,"posy":258,"posx":-12.5},"28":{"pot_scale":1,"posy":66.5,"posx":-11},"27":{"pot_scale":1,"posy":238.5,"posx":-26},"26":{"pot_scale":1,"posy":73.5,"posx":-7},"16":{"pot_scale":1,"posy":62,"posx":-9.5},"19":{"pot_scale":1,"posy":238.5,"posx":-26},"18":{"pot_scale":1,"posy":73.5,"posx":-7},"1":{"posy":-29.5,"posx":-22.5},"2":{"posy":239.5,"posx":-11.5},"3":{"posy":128.5,"posx":-21},"4":{"posy":253.5,"posx":-11.5},"5":{"posy":56,"posx":-9.5},"13":{"posy":253.5,"posx":-11.5},"12":{"posy":205.5,"posx":-24},"15":{"pot_scale":1,"posy":257.5,"posx":-18},"14":{"posy":56,"posx":-9.5},"10":{"posy":-204.5,"posx":-9.5},"11":{"posy":417.5,"posx":-21.5},"21":{"pot_scale":1,"posy":258,"posx":-12.5},"20":{"pot_scale":1,"posy":66.5,"posx":-11}},"id":1}