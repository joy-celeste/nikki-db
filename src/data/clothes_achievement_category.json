{"8":"Troupe","9":"4 Seasons","5":"Others","4":"Styling Contest","7":"Festivals","6":"7 Nations","3":"Journey","21":"Museum","20":"御苑琼芳","23":"Constellation Hall","22":"Ancient Fossil Hall","15":"North","14":"Pigeon","17":"Ruin","16":"Wasteland","18":"Happiness","19":"Story Suit","2":"Collection","1":"My Growth","12":"Lilith","13":"Cloud","10":"Stars","11":"Apple"}