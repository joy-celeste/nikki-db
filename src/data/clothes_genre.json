{"Coat":3,"AccessoryCount":8,"AccessoryEnd":17,"Earring":11,"Shoes":7,"Necklace":12,"Count":10,"Top":4,"Special":16,"Bottom":5,"Hair":1,"Dress":2,"Headwear":10,"Accessory":8,"AccessoryStart":10,"Waistdeco":15,"Bracelet":13,"Socks":6,"Skin":17,"MiracleSoul":88,"Makeup":9,"Handheld":14}