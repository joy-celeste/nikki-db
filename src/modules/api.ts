import { HttpRequestException } from './errors';
import database from './database';
import { API_CONSTANTS } from './constants';

export const fetchItemData = async (itemId: number): Promise<any> =>
  database.collection(API_CONSTANTS.CLOTHES).doc(itemId.toString())
    .get()
    .then((response) => {
      if (!response) {
        console.log('No such document!');
        return null;
      }
      console.log('Retrieved data from database.');
      return response.data();
    })
    .catch((error) => {
      console.log(new HttpRequestException(error, 'Error getting response.'));
    });

export {};
