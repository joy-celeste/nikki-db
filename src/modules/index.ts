import { combineReducers } from 'redux';
import { userReducer } from './user';
import { productsReducer } from './products';

export const rootReducer = combineReducers({
  user: userReducer,
  products: productsReducer,
});

export type RootState = ReturnType<typeof rootReducer>;