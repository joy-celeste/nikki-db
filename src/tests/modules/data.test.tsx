import { Store } from 'redux';
import { posedDress, posedCoat, posedShoes, simpleHair,
  simpleHat, complexHair, deerSpirit, motorcycleSpirit, simpleDress } from '../test_data/data';
import { DeserializeNullException } from '../../modules/errors';
import { RootState } from '../../modules';
import { ItemData, loadItem, loadMultipleItems, PositionData, initialState } from '../../modules/data';
import { createStoreWithMiddleware } from '../helpers';
import * as api from '../../modules/api';

const mockMath = Object.create(global.Math);
mockMath.random = () => 0.01;
global.Math = mockMath;

describe('ItemData', () => {
  test('Errors if given empty data', () => {
    expect(() => new ItemData()).toThrowError(DeserializeNullException);
    expect(() => new PositionData(0, null)).toThrowError(DeserializeNullException);
  });

  test('Successfully deserializes for the null cases', () => {
    const emptyItemData: ItemData = new ItemData({
    });
    expect(emptyItemData).toMatchSnapshot();
    const emptyPositionData: PositionData = new PositionData(10, {
    });
    expect(emptyPositionData).toMatchSnapshot();
  });

  test('Successfully deserializes ItemData: simpleHair', () => {
    const simpleHairItem: ItemData = new ItemData(simpleHair);
    expect(simpleHairItem).toMatchSnapshot();
    expect(simpleHairItem.amputationData).toBe(null);
  });

  test('Successfully deserializes ItemData: posed dress', () => {
    const posedDressItem: ItemData = new ItemData(posedDress);
    expect(posedDressItem).toMatchSnapshot();
  });

  test('Successfully deserializes ItemData: posed coat', () => {
    const posedCoatItem: ItemData = new ItemData(posedCoat);
    expect(posedCoatItem).toMatchSnapshot();
  });

  test('Successfully deserializes ItemData: posed shoes', () => {
    const posedShoesItem: ItemData = new ItemData(posedShoes);
    expect(posedShoesItem).toMatchSnapshot();
  });

  test('Successfully deserializes ItemData: simple hat', () => {
    const simpleHatItem: ItemData = new ItemData(simpleHat);
    expect(simpleHatItem).toMatchSnapshot();
  });

  test('Successfully deserializes ItemData: complex hair', () => {
    const complexHairItem: ItemData = new ItemData(complexHair);
    expect(complexHairItem).toMatchSnapshot();
  });

  test('Successfully deserializes ItemData: motorcycle spririt', () => {
    const motorcycleSpiritItem: ItemData = new ItemData(motorcycleSpirit);
    expect(motorcycleSpiritItem).toMatchSnapshot();
  });

  test('Successfully deserializes ItemData: deer spirit', () => {
    const deerSpiritItem: ItemData = new ItemData(deerSpirit);
    expect(deerSpiritItem).toMatchSnapshot();
  });
});

describe('DataState - with local clothesData', () => {
  let store: Store<any>;

  beforeEach(() => {
    store = createStoreWithMiddleware();
  });

  test('Action: ADD_ITEMS / Use-case: loadItem -- document does not exist', async () => {
    await store.dispatch<any>(loadItem(999999));
    const state: RootState = store.getState();
    expect(state.data.itemsData).toStrictEqual(initialState.itemsData);
    expect(state.data.itemsData[posedDress.id]).toStrictEqual(undefined);
  });

  test('Action: ADD_ITEMS / Use-case: loadItem -- successfully calls looks up item and adds it to the store', async () => {
    await store.dispatch<any>(loadItem(posedDress.id));
    const state: RootState = store.getState();
    expect(state.data.itemsData[posedDress.id]).toStrictEqual(new ItemData(posedDress));
  });

  test('Action: ADD_ITEMS / Use-case: loadItem -- successfully adds multiple items to the store', async () => {
    await store.dispatch<any>(loadItem(posedCoat.id));
    await store.dispatch<any>(loadItem(posedShoes.id));
    const state: RootState = store.getState();
    expect(Object.keys(state.data.itemsData)).toHaveLength(3); // nikki's pinky, coat, shoes
    expect(state.data.itemsData[posedCoat.id]).toStrictEqual(new ItemData(posedCoat));
    expect(state.data.itemsData[posedShoes.id]).toStrictEqual(new ItemData(posedShoes));
    expect(state.data.itemsData).toMatchSnapshot();
  });

  test(`Action: ADD_ITEMS / Use-case: loadMultipleItems - 
        load multiple items successfully (no conflicts)`, async () => {
    let state: RootState;
    state = store.getState();
    expect(state.character.history.length).toEqual(1);
    expect(state.character.step).toEqual(0);
    expect(Object.keys(state.character.history[0].clothes).length).toEqual(1);
    expect(Object.keys(state.data.itemsData)).toHaveLength(1); // nikki's pinky

    await store.dispatch<any>(loadMultipleItems([posedCoat.id, posedDress.id]));
    state = store.getState();
    expect(state.character.history.length).toEqual(2);
    expect(state.character.step).toEqual(1);
    expect(Object.keys(state.character.history[1].clothes).length).toEqual(3);
    expect(Object.keys(state.data.itemsData)).toHaveLength(3); // nikki's pinky, coat, dress

    expect(state.data.itemsData[posedCoat.id]).toStrictEqual(new ItemData(posedCoat));
    expect(state.data.itemsData[posedDress.id]).toStrictEqual(new ItemData(posedDress));
    expect(state.data.itemsData).toMatchSnapshot();
  });

  test(`Action: ADD_ITEMS / Use-case: loadMultipleItems - load multiple items successfully (one conflict)`, async () => {
    await store.dispatch<any>(loadMultipleItems([posedCoat.id, simpleHair.id]));
    const state: RootState = store.getState();
    expect(state.character.history.length).toEqual(2);
    expect(state.character.step).toEqual(1);
    expect(Object.keys(state.character.history[1].clothes).length).toEqual(2);
    expect(Object.keys(state.data.itemsData)).toHaveLength(3); // nikki's pinky, coat, hair
    expect(state.data.itemsData[posedCoat.id]).toStrictEqual(new ItemData(posedCoat));
    expect(state.data.itemsData[simpleHair.id]).toStrictEqual(new ItemData(simpleHair));
    expect(state.data.itemsData).toMatchSnapshot();
  });

  test('Action: ADD_ITEMS / Use-case: loadMultipleItems - use item data if already loaded', async () => {
    let state: RootState;

    await store.dispatch<any>(loadMultipleItems([posedCoat.id, posedDress.id, simpleHair.id]));
    state = store.getState();
    expect(state.character.history.length).toEqual(2);
    expect(state.character.step).toEqual(1);
    expect(Object.keys(state.character.history[1].clothes).length).toEqual(3); // nikki's pinky, hair, dress, coat
    expect(Object.keys(state.data.itemsData)).toHaveLength(4);
    expect(state.data).toMatchSnapshot();
    expect(state.character).toMatchSnapshot();

    await store.dispatch<any>(loadMultipleItems([posedCoat.id, simpleDress.id, complexHair.id]));
    state = store.getState();
    expect(state.character.history.length).toEqual(3);
    expect(state.character.step).toEqual(2);
    expect(Object.keys(state.character.history[2].clothes).length).toEqual(3);
    expect(Object.keys(state.data.itemsData)).toHaveLength(6);
    expect(state.data).toMatchSnapshot();
    expect(state.character).toMatchSnapshot();
  });

  test(`Action: ADD_ITEMS / Use-case: loadMultipleItems - handle API call returning blank data`, async () => {
    await store.dispatch<any>(loadMultipleItems([posedCoat.id, posedDress.id]));
    const state: RootState = store.getState();
    expect(state.character.history.length).toEqual(2);
    expect(state.character.step).toEqual(1);
    expect(Object.keys(state.character.history[1].clothes).length).toEqual(3);
    expect(Object.keys(state.data.itemsData)).toHaveLength(3);
    expect(state.data).toMatchSnapshot();
    expect(state.character).toMatchSnapshot();
  });

  test(`Action: ADD_ITEMS / Use-case: loadMultipleItems - 
        handle rejections in API calls - load as many items as possible`, async () => {
    await store.dispatch<any>(loadMultipleItems([posedCoat.id, posedDress.id, `invalidId`]));
    const state: RootState = store.getState();
    expect(state.character.history.length).toEqual(2);
    expect(state.character.step).toEqual(1);
    expect(Object.keys(state.character.history[1].clothes).length).toEqual(3);
    expect(Object.keys(state.data.itemsData)).toHaveLength(3);
    expect(state.data).toMatchSnapshot();
    expect(state.character).toMatchSnapshot();
  });
});

describe.skip('DataState - with API', () => {
  let store: Store<any>;
  let apiMock: any;

  beforeEach(() => {
    store = createStoreWithMiddleware();
    apiMock = jest.spyOn(api, 'fetchItemData');
    apiMock.mockClear();
  });

  test('Action: ADD_ITEMS / Use-case: loadItem -- document does not exist', async () => {
    apiMock.mockImplementation(); // Document does not exist
    await store.dispatch<any>(loadItem(posedDress.id));
    const state: RootState = store.getState();

    expect(apiMock).toBeCalledTimes(1);
    expect(state.data.itemsData).toStrictEqual({
    });
    expect(state.data.itemsData[posedDress.id]).toStrictEqual(undefined);
  });

  test('Action: ADD_ITEMS / Use-case: loadItem -- successfully calls API and adds one item to the store', async () => {
    apiMock.mockImplementation(() => posedDress);
    await store.dispatch<any>(loadItem(posedDress.id));
    const state: RootState = store.getState();

    expect(apiMock).toBeCalledTimes(1);
    expect(state.data.itemsData[posedDress.id]).toStrictEqual(new ItemData(posedDress));
  });

  test('Action: ADD_ITEMS / Use-case: loadItem -- successfully adds multiple items to the store', async () => {
    apiMock.mockImplementation(() => posedCoat);
    await store.dispatch<any>(loadItem(posedCoat.id));

    apiMock.mockImplementation(() => posedShoes);
    await store.dispatch<any>(loadItem(posedShoes.id));
    const state: RootState = store.getState();

    expect(apiMock).toBeCalledTimes(2);
    expect(Object.keys(state.data.itemsData)).toHaveLength(2);
    expect(state.data.itemsData[posedCoat.id]).toStrictEqual(new ItemData(posedCoat));
    expect(state.data.itemsData[posedShoes.id]).toStrictEqual(new ItemData(posedShoes));
    expect(state.data.itemsData).toMatchSnapshot();
  });

  test('Action: ADD_ITEMS / Use-case: loadItem -- calls API only once if the item is already loaded', async () => {
    apiMock.mockImplementation(() => posedCoat);
    await store.dispatch<any>(loadItem(posedCoat.id));

    apiMock.mockImplementation(() => posedCoat);
    await store.dispatch<any>(loadItem(posedCoat.id));
    const state: RootState = store.getState();

    expect(apiMock).toBeCalledTimes(1);
    expect(Object.keys(state.data.itemsData)).toHaveLength(1);
    expect(state.data.itemsData[posedCoat.id]).toStrictEqual(new ItemData(posedCoat));
    expect(state.data.itemsData).toMatchSnapshot();
  });

  test(`Action: ADD_ITEMS / Use-case: loadMultipleItems - 
        load multiple items successfully (no conflicts)`, async () => {
    let state: RootState;

    apiMock
      .mockImplementationOnce(() => posedCoat)
      .mockImplementationOnce(() => posedDress);

    state = store.getState();
    expect(state.character.history.length).toEqual(1);
    expect(state.character.step).toEqual(0);
    expect(Object.keys(state.character.history[0].clothes).length).toEqual(1);
    expect(Object.keys(state.data.itemsData)).toHaveLength(0);

    await store.dispatch<any>(loadMultipleItems([posedCoat.id, posedDress.id]));
    state = store.getState();
    expect(apiMock).toBeCalledTimes(2);
    expect(state.character.history.length).toEqual(2);
    expect(state.character.step).toEqual(1);
    expect(Object.keys(state.character.history[1].clothes).length).toEqual(3);
    expect(Object.keys(state.data.itemsData)).toHaveLength(2);

    expect(state.data.itemsData[posedCoat.id]).toStrictEqual(new ItemData(posedCoat));
    expect(state.data.itemsData[posedDress.id]).toStrictEqual(new ItemData(posedDress));
    expect(state.data.itemsData).toMatchSnapshot();
  });

  test(`Action: ADD_ITEMS / Use-case: loadMultipleItems -
        load multiple items successfully (one conflict)`, async () => {
    apiMock
      .mockImplementationOnce(() => posedCoat)
      .mockImplementationOnce(() => simpleHair);

    await store.dispatch<any>(loadMultipleItems([posedCoat.id, simpleHair.id]));
    const state: RootState = store.getState();
    expect(apiMock).toBeCalledTimes(2);
    expect(state.character.history.length).toEqual(2);
    expect(state.character.step).toEqual(1);
    expect(Object.keys(state.character.history[1].clothes).length).toEqual(2);
    expect(Object.keys(state.data.itemsData)).toHaveLength(2);

    expect(state.data.itemsData[posedCoat.id]).toStrictEqual(new ItemData(posedCoat));
    expect(state.data.itemsData[simpleHair.id]).toStrictEqual(new ItemData(simpleHair));
    expect(state.data.itemsData).toMatchSnapshot();
  });

  test('Action: ADD_ITEMS / Use-case: loadMultipleItems - use item data if already loaded', async () => {
    let state: RootState;

    apiMock
      .mockImplementationOnce(() => posedCoat)
      .mockImplementationOnce(() => posedDress)
      .mockImplementationOnce(() => simpleHair)
      .mockImplementationOnce(() => simpleDress)
      .mockImplementationOnce(() => complexHair);

    await store.dispatch<any>(loadMultipleItems([posedCoat.id, posedDress.id, simpleHair.id]));
    state = store.getState();
    expect(apiMock).toBeCalledTimes(3);
    expect(state.character.history.length).toEqual(2);
    expect(state.character.step).toEqual(1);
    expect(Object.keys(state.character.history[1].clothes).length).toEqual(3);
    expect(Object.keys(state.data.itemsData)).toHaveLength(3);
    expect(state.data).toMatchSnapshot();
    expect(state.character).toMatchSnapshot();

    await store.dispatch<any>(loadMultipleItems([posedCoat.id, simpleDress.id, complexHair.id]));
    state = store.getState();
    expect(apiMock).toBeCalledTimes(5);
    expect(state.character.history.length).toEqual(3);
    expect(state.character.step).toEqual(2);
    expect(Object.keys(state.character.history[2].clothes).length).toEqual(3);
    expect(Object.keys(state.data.itemsData)).toHaveLength(5);
    expect(state.data).toMatchSnapshot();
    expect(state.character).toMatchSnapshot();
  });

  test(`Action: ADD_ITEMS / Use-case: loadMultipleItems - 
        handle API call returning blank data`, async () => {
    apiMock
      .mockImplementationOnce(() => posedCoat)
      .mockImplementationOnce(() => {});

    await store.dispatch<any>(loadMultipleItems([posedCoat.id, posedDress.id]));
    const state: RootState = store.getState();
    expect(apiMock).toBeCalledTimes(2);
    expect(state.character.history.length).toEqual(2);
    expect(state.character.step).toEqual(1);
    expect(Object.keys(state.character.history[1].clothes).length).toEqual(2);
    expect(Object.keys(state.data.itemsData)).toHaveLength(1);
    expect(state.data).toMatchSnapshot();
    expect(state.character).toMatchSnapshot();
  });

  test(`Action: ADD_ITEMS / Use-case: loadMultipleItems - 
        handle rejections in API calls - load as many items as possible`, async () => {
    apiMock
      .mockImplementationOnce(() => posedCoat)
      .mockImplementationOnce(() => posedDress)
      .mockImplementationOnce(() => Promise.reject(new Error('oops')));

    await store.dispatch<any>(loadMultipleItems([posedCoat.id, posedDress.id, posedShoes.id]));
    const state: RootState = store.getState();
    expect(apiMock).toBeCalledTimes(3);
    expect(state.character.history.length).toEqual(2);
    expect(state.character.step).toEqual(1);
    expect(Object.keys(state.character.history[1].clothes).length).toEqual(3);
    expect(Object.keys(state.data.itemsData)).toHaveLength(2);
    expect(state.data).toMatchSnapshot();
    expect(state.character).toMatchSnapshot();
  });

  
});
