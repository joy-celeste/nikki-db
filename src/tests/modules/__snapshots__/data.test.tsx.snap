// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataState Action: ADD_ITEMS / Use-case: loadItem -- calls API only once if the item is already loaded 1`] = `
Object {
  "30987": ItemData {
    "amputationData": Object {
      "10": false,
      "12": false,
      "5": false,
      "9": true,
    },
    "depthType": 48,
    "depths": Object {
      "1": 36700,
      "12": 10190,
      "13": 10160,
      "14": 9880,
      "2": 27000,
      "3": 19000,
      "4": 5000,
      "5": 840,
    },
    "description": "He stands against the light and looks back at her. She saves the throb deeply into her heart.",
    "id": 30987,
    "name": "Rose Heart",
    "position": Array [
      PositionData {
        "depth": 1,
        "scale": 1,
        "x": -27,
        "y": 200.5,
        "z": 36700,
      },
      PositionData {
        "depth": 2,
        "scale": 1,
        "x": -202,
        "y": -130,
        "z": 27000,
      },
      PositionData {
        "depth": 3,
        "scale": 1,
        "x": 101,
        "y": -119,
        "z": 19000,
      },
      PositionData {
        "depth": 4,
        "scale": 1,
        "x": -81,
        "y": -153,
        "z": 5000,
      },
      PositionData {
        "depth": 13,
        "scale": 1,
        "x": -88.5,
        "y": 245,
        "z": 10160,
      },
      PositionData {
        "depth": 14,
        "scale": 1,
        "x": 42.5,
        "y": 234.5,
        "z": 9880,
      },
      PositionData {
        "depth": 101,
        "scale": 1,
        "x": -30.5,
        "y": 242.5,
        "z": 27000,
      },
      PositionData {
        "depth": 201,
        "scale": 1,
        "x": -47.5,
        "y": 262,
        "z": 19000,
      },
      PositionData {
        "depth": 301,
        "scale": 1,
        "x": 124.5,
        "y": 139,
        "z": 5000,
      },
    ],
    "subType": 3,
  },
}
`;

exports[`DataState Action: ADD_ITEMS / Use-case: loadItem -- successfully adds multiple items to the store 1`] = `
Object {
  "30987": ItemData {
    "amputationData": Object {
      "10": false,
      "12": false,
      "5": false,
      "9": true,
    },
    "depthType": 48,
    "depths": Object {
      "1": 36700,
      "12": 10190,
      "13": 10160,
      "14": 9880,
      "2": 27000,
      "3": 19000,
      "4": 5000,
      "5": 840,
    },
    "description": "He stands against the light and looks back at her. She saves the throb deeply into her heart.",
    "id": 30987,
    "name": "Rose Heart",
    "position": Array [
      PositionData {
        "depth": 1,
        "scale": 1,
        "x": -27,
        "y": 200.5,
        "z": 36700,
      },
      PositionData {
        "depth": 2,
        "scale": 1,
        "x": -202,
        "y": -130,
        "z": 27000,
      },
      PositionData {
        "depth": 3,
        "scale": 1,
        "x": 101,
        "y": -119,
        "z": 19000,
      },
      PositionData {
        "depth": 4,
        "scale": 1,
        "x": -81,
        "y": -153,
        "z": 5000,
      },
      PositionData {
        "depth": 13,
        "scale": 1,
        "x": -88.5,
        "y": 245,
        "z": 10160,
      },
      PositionData {
        "depth": 14,
        "scale": 1,
        "x": 42.5,
        "y": 234.5,
        "z": 9880,
      },
      PositionData {
        "depth": 101,
        "scale": 1,
        "x": -30.5,
        "y": 242.5,
        "z": 27000,
      },
      PositionData {
        "depth": 201,
        "scale": 1,
        "x": -47.5,
        "y": 262,
        "z": 19000,
      },
      PositionData {
        "depth": 301,
        "scale": 1,
        "x": 124.5,
        "y": 139,
        "z": 5000,
      },
    ],
    "subType": 3,
  },
  "71927": ItemData {
    "amputationData": Object {
      "10": true,
      "12": false,
      "5": true,
      "9": false,
    },
    "depthType": 16,
    "depths": Object {
      "1": 13750,
      "2": 13500,
      "3": 9000,
      "9": 10220,
    },
    "description": "Those old stories fade into dust and get blown away by the wind. Yi walks in the mortal world and listens.",
    "id": 71927,
    "name": "Sparse Stars",
    "position": Array [
      PositionData {
        "depth": 1,
        "scale": 1,
        "x": -365.5,
        "y": -487,
        "z": 13750,
      },
      PositionData {
        "depth": 4,
        "scale": 1,
        "x": -672.5,
        "y": -357,
        "z": undefined,
      },
    ],
    "subType": 8,
  },
}
`;

exports[`ItemData Successfully deserializes ItemData: complex hair 1`] = `
ItemData {
  "amputationData": null,
  "depthType": 1,
  "depths": Object {
    "1": 33000,
    "2": 3500,
    "7": 38000,
    "8": 37200,
    "9": 3510,
  },
  "description": "As the night falls, the grand theatre is all empty. The touching music behind the curtain feels strangely beautiful.",
  "id": 12363,
  "name": "Dreamland Snow",
  "position": Array [
    PositionData {
      "depth": 1,
      "scale": 1,
      "x": -25,
      "y": 378,
      "z": 33000,
    },
    PositionData {
      "depth": 2,
      "scale": 1,
      "x": 254.5,
      "y": 45.5,
      "z": 3500,
    },
    PositionData {
      "depth": 9,
      "scale": 1,
      "x": 81,
      "y": 55.5,
      "z": 3510,
    },
  ],
  "subType": 1,
}
`;

exports[`ItemData Successfully deserializes ItemData: deer spirit 1`] = `
ItemData {
  "amputationData": null,
  "depthType": 61,
  "depths": Object {
    "1": 60000,
  },
  "description": "The season may change, but flowers will bloom if there is spring in your heart. The boy never forgets this place.",
  "id": 880048,
  "name": "Deer and Cliff",
  "position": Array [
    PositionData {
      "depth": 1,
      "scale": null,
      "x": 124.5,
      "y": 51,
      "z": 60000,
    },
  ],
  "subType": 34,
}
`;

exports[`ItemData Successfully deserializes ItemData: motorcycle spririt 1`] = `
ItemData {
  "amputationData": null,
  "depthType": 61,
  "depths": Object {
    "1": 60000,
  },
  "description": "At 24, Louie wore the white robe of Nameless Order and held the knight spear, leaving Bluebird and the past behind",
  "id": 880092,
  "name": "Dark Night Motor",
  "position": Array [
    PositionData {
      "depth": 1,
      "scale": null,
      "x": 154,
      "y": -200,
      "z": 60000,
    },
  ],
  "subType": 34,
}
`;

exports[`ItemData Successfully deserializes ItemData: posed coat 1`] = `
ItemData {
  "amputationData": Object {
    "10": false,
    "12": false,
    "5": false,
    "9": true,
  },
  "depthType": 48,
  "depths": Object {
    "1": 36700,
    "12": 10190,
    "13": 10160,
    "14": 9880,
    "2": 27000,
    "3": 19000,
    "4": 5000,
    "5": 840,
  },
  "description": "He stands against the light and looks back at her. She saves the throb deeply into her heart.",
  "id": 30987,
  "name": "Rose Heart",
  "position": Array [
    PositionData {
      "depth": 1,
      "scale": 1,
      "x": -27,
      "y": 200.5,
      "z": 36700,
    },
    PositionData {
      "depth": 2,
      "scale": 1,
      "x": -202,
      "y": -130,
      "z": 27000,
    },
    PositionData {
      "depth": 3,
      "scale": 1,
      "x": 101,
      "y": -119,
      "z": 19000,
    },
    PositionData {
      "depth": 4,
      "scale": 1,
      "x": -81,
      "y": -153,
      "z": 5000,
    },
    PositionData {
      "depth": 13,
      "scale": 1,
      "x": -88.5,
      "y": 245,
      "z": 10160,
    },
    PositionData {
      "depth": 14,
      "scale": 1,
      "x": 42.5,
      "y": 234.5,
      "z": 9880,
    },
    PositionData {
      "depth": 101,
      "scale": 1,
      "x": -30.5,
      "y": 242.5,
      "z": 27000,
    },
    PositionData {
      "depth": 201,
      "scale": 1,
      "x": -47.5,
      "y": 262,
      "z": 19000,
    },
    PositionData {
      "depth": 301,
      "scale": 1,
      "x": 124.5,
      "y": 139,
      "z": 5000,
    },
  ],
  "subType": 3,
}
`;

exports[`ItemData Successfully deserializes ItemData: posed dress 1`] = `
ItemData {
  "amputationData": Object {
    "10": true,
    "12": true,
    "5": false,
    "9": true,
  },
  "depthType": 41,
  "depths": Object {
    "1": 36600,
    "12": 10200,
    "13": 10170,
    "14": 9900,
    "2": 24000,
    "21": 24000,
    "22": 24000,
    "3": 18000,
    "4": 6000,
    "5": 850,
    "8": 27150,
    "9": 10220,
  },
  "description": "The sakura festival has come again. In the flower rain, she holds an umbrella, waiting for the one to come.",
  "id": 22008,
  "name": "Sakura Dream",
  "position": Array [
    PositionData {
      "depth": 2,
      "scale": 1,
      "x": -32.5,
      "y": -164.5,
      "z": 24000,
    },
    PositionData {
      "depth": 901,
      "scale": 1,
      "x": -59.5,
      "y": 328,
      "z": undefined,
    },
  ],
  "subType": 2,
}
`;

exports[`ItemData Successfully deserializes ItemData: posed shoes 1`] = `
ItemData {
  "amputationData": Object {
    "10": true,
    "12": false,
    "5": true,
    "9": false,
  },
  "depthType": 16,
  "depths": Object {
    "1": 13750,
    "2": 13500,
    "3": 9000,
    "9": 10220,
  },
  "description": "Those old stories fade into dust and get blown away by the wind. Yi walks in the mortal world and listens.",
  "id": 71927,
  "name": "Sparse Stars",
  "position": Array [
    PositionData {
      "depth": 1,
      "scale": 1,
      "x": -365.5,
      "y": -487,
      "z": 13750,
    },
    PositionData {
      "depth": 4,
      "scale": 1,
      "x": -672.5,
      "y": -357,
      "z": undefined,
    },
  ],
  "subType": 8,
}
`;

exports[`ItemData Successfully deserializes ItemData: simple hat 1`] = `
ItemData {
  "amputationData": null,
  "depthType": 19,
  "depths": Object {
    "1": 36000,
    "2": 3000,
    "6": 35150,
    "7": 890,
    "8": 37300,
  },
  "description": "Every cowboy in the Apple Federal has a cap like this.",
  "id": 80078,
  "name": "Denim CapÂ·Rare",
  "position": Array [
    PositionData {
      "depth": 1,
      "scale": null,
      "x": -17.5,
      "y": 426,
      "z": 36000,
    },
    PositionData {
      "depth": 2,
      "scale": null,
      "x": -96.5,
      "y": 447,
      "z": 3000,
    },
  ],
  "subType": 10,
}
`;

exports[`ItemData Successfully deserializes ItemData: simpleHair 1`] = `
ItemData {
  "amputationData": null,
  "depthType": 1,
  "depths": Object {
    "1": 33000,
    "2": 3500,
    "7": 38000,
    "8": 37200,
    "9": 3510,
  },
  "description": "A gentle, thoughtful nobleman is not supposed to be the silent background forever.",
  "id": 10007,
  "name": "Elegant Nobleman",
  "position": Array [
    PositionData {
      "depth": 1,
      "scale": null,
      "x": -21.5,
      "y": 426,
      "z": 33000,
    },
  ],
  "subType": 1,
}
`;

exports[`ItemData Successfully deserializes for the null cases 1`] = `
ItemData {
  "amputationData": null,
  "depthType": null,
  "depths": null,
  "description": "",
  "id": null,
  "name": "",
  "position": null,
  "subType": null,
}
`;

exports[`ItemData Successfully deserializes for the null cases 2`] = `
PositionData {
  "depth": 10,
  "scale": null,
  "x": undefined,
  "y": undefined,
  "z": null,
}
`;
